woocommerce/woocommerce-android
# Feat/products 2707 simple product details add base

## RESOURCES
[The Figma file](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=460%3A1760&viewport=-1704%2C1577%2C0.5334523320198059) and [the initial issue](https://github.com/woocommerce/woocommerce-android/issues/2707)

## CHANGES / DESCRIPTION
This PR will contain the following:
- added the ability of automatically recording a default product type selection, upon adding a new product for the first time (as seen [here](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=80%3A0&viewport=-4154%2C2459%2C1.013650894165039));
- added the **base/default** product detail screen for an "add new product flow".

**1)** In order to have the user preference recorded for a product type, I used the `AppPrefs` as it seemed the best place to do so - it already contained some user preferences. I did however try to hide the implementation of what a "preference mechanism" is in this system by making the call explicitly from the `viewModel`; with a bit more detail:
 - the call to `setSelectedProductType` in the `ProductTypesBottomSheetViewModel`;
 - the call to `getSelectedProductType` in the `ProductListViewModel`.

In the future, if we change `AppPrefs` to a more complex system (say a local DB), the UI side (activity, views, adapters, etc) won't need to be updated to handle this new implementation detail and should work as normal 👍 

**2)** The second part of this PR is the actual **Product Add** screen. 

My aim was to reuse the most I could from the `ProductDetailFragment` and to get the upcoming product steps (image, details, price, etc) as it is expected from a product detail like screen. In order to offer a split of implementation, for an **"edit"** or an **"add"** product flow, I added a new `argument` (just as in [this other PR](https://github.com/woocommerce/woocommerce-android/pull/2720)) to allow me to easily split the paths in the common `ProductDetailViewModel`; the `start()` method of the view model will dictate the path to take.

For the "**add** new product" path, I also added a new view state parameter, called`isAddNewProduct`. This parameter, will be part of the setup of observers in the `ProductDetailFragment` and like so, trigger the correct UI visibility for this scenario.

I also added the new top nav bar with a default name of `New Product` 👌  speaking of the top nav bar, there's a small caveat that it doesn't seem to like my approach for the `menu_done` and it is not showing the **PUBLISH** item - but I will look into it, it's not a massive priority right now.

As always, let me know your thoughts on the matter 🙏 

## TESTING
- Click on the PRODUCT bottom nav bar button
    - Wait for the screen to finish loading
 - Scenario 1:
    - Click on the FAB button for the first time, on a fresh install
    - Notice the bottom sheet for `Product Add types` appearing with the product list showing behind
    - Select a product type
    - Notice the product detail for a new product flow, empty with some default views
 - Scenario 2
   - Click on the FAB button not for the first time
   - Notice the product detail screen for a new product showing straight away, empty with some default views

## SCREENSHOTS
<image src="https://user-images.githubusercontent.com/7058393/90453744-5d90e480-e0e9-11ea-9c41-a5df90b48f86.gif" width="320"/>
<image src="https://user-images.githubusercontent.com/7058393/90453743-5c5fb780-e0e9-11ea-88c9-d6ee38c07178.gif" width="320"/>
